llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -8883,6 +8883,29 @@ static std::optional<bool> isImpliedCondICmps(const ICmpInst *LHS,
   if (L0 == R0 && L1 == R1)
     return isImpliedCondMatchingOperands(LPred, RPred);
 
+  // Take SGT as an example:  L0:x > L1:y and C >= 0
+  //                      ==> R0:(x -nsw y) < R1:(-C) is false
+  Value *X, *Y;
+  if (((LPred == ICmpInst::ICMP_SGT || LPred == ICmpInst::ICMP_SGE) &&
+       match(R0, m_NSWSub(m_Value(X), m_Value(Y)))) ||
+      ((LPred == ICmpInst::ICMP_UGT || LPred == ICmpInst::ICMP_UGE) &&
+       match(R0, m_NUWSub(m_Value(X), m_Value(Y))))) {
```

Comment:
Here and below, don't think you need `nuw` for the sub. Its implied by LPred.

Reply:
https://alive2.llvm.org/ce/z/yQTr5f

Reply:
Actually I misspoke, but I don't think your proofs are correct. The `nuw` on the `sub nuw i8 0, %add` forces `%add` to be 0. Likewise the `sub nuw i8 %sub, %c` essentially forces the condition you want, not the dominating predicates.

Reply:
Thanks @goldsteinn , If this only valid for 0 with nuw, so the transformation of unsigned should be deleted (only keep it for signed types) ? or do you have other suggestion ？

Reply:
Yeah, unsigned doesn't make sense here. The is really saying "X - Y must be positive if X >= Y and no overflow". That only really makes sense in the context of signed comparison.

Reply:
Thanks for the confirmation, and delete the parts of unsigned.